design/configuration-crd.md

```diff
@@ -0,0 +1,210 @@
+    configmapName: leader-elect
+    configmapNamespace: projectcontour
+    disableLeaderElection: false
+  enableExternalNameService: false
+  network:
+    numTrustedHops: 0
+    adminPort: 9001
+  rateLimitService:
+    extensionService: projectcontour/ratelimit
+    domain: contour
+    failOpen: false
+    enableXRateLimitHeaders: false
+  policy:
+    requestHeaders:
+      set:
+    responseHeaders:
+      set:
+status:
+```
+
+## Converting from Configmap
+
+Contour will provide a way internally to move to the new CRD and not require users to manually migrate to the new CRD format.
+Contour will provide a new command or external tool (similar to ir2proxy) which will migrate between the specs accordingly. 
+
+## Contour Deployment
+
+A managed version of Contour was made available with the `Contour Operator`.
+Since Contour will manage Envoy instances, the Operator will now manage instances of Contour.
+The details of how an instance of Contour should be deployed within a cluster will be defined in the second CRD named `ContourDeployment`. 
+The `spec.confguration` of this object will be the same struct defined in the `ContourConfiguration`. 
+
+A controller will watch for these objects to be created and take action on them accordingly to make desired state in the cluster match the configuration on the spec. 
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ContourDeployment
+metadata:
+  name: contour
+spec:
+  replicas: 2
+  nodePlacement:
+    nodeSelector:
+    tolerations:
+  configuration:
+    <same config as above>
+status:
+```
+
+## Processing Logic
+
+Contour will require a new flag (`--contour-config`), which will allow for customizing the name of the `ContourConfiguration` CRD that is it to use.
+It will default to one named `contour`, but could be overridden if desired.
+The ContourConfiguration referenced must also be in the same namespace as Contour is running, it's not valid to reference a configuration from another namespace.
+The current flag `--config-path`/`-c` will continue to point to the Configmap file, but over time could eventually be deprecated and the short code `-c` be used for the CRD location (i.e. `--contour-config`) for simplicity.
+The Contour configuration CRD will still remain optional.
+In its absence, Contour will operate with reasonable defaults.
+Where Contour settings can also be specified with command-line flags, the command-line value takes precedence over the configuration file.
+
+On startup, Contour will look for a `ContourConfiguration` CRD named `contour` or using the value of the flag `--contour-config` in the same namespace which Contour is running in, Contour won't support referencing from a different namespace.
+If the `ContourConfiguration` CRD is not found Contour will start up with reasonable defaults.
+
+Contour will set status on the object informing the user if there are any errors or issues with the config file or that is all fine and processing correctly.
```

Comment:
I just updated the text a bit, but my thought was to make Contour not ready, but wait for a valid config. While not healthy it won't serve any envoy's configuration.

Reply:
I think that idea sounds good, but it's harder to catch the error then. Many clusters have alerts or other measures to detect crashloops, whereas this would require special handling. I'd rather keep crashloop until we have something that handles failure better.

Reply:
> until we have something that handles failure better.

What would that look like? A pod going "NotReady" is a pretty core concept in k8s to indicate a problem.